[shared_preferences] Migrate main package to null-safety

gaetschwartz opened the pull request:
Migrates `shared_preferences` to nnbd.

Fixes https://github.com/flutter/flutter/issues/74876.

**Note:** also recreated the android and macos `example` applications because the macOS app was apparently a copy of the connectivity app and had `connectivity_example` as its name, the Android app was using the embedding v1. 

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

stuartmorgan commented on packages/shared_preferences/shared_preferences/pubspec.yaml:
```diff
@@ -26,16 +26,16 @@ dependencies:
+  shared_preferences_macos: ^0.0.2-nullsafety
+  shared_preferences_web: ^0.2.0-nullsafety
+  shared_preferences_windows: ^0.0.3-nullsafety
 
```
You didn't change the SDK min version, so there's no actual migration here.

stuartmorgan reviewed (changes requested):
This doesn't actually do a null-safety migration; there's no SDK version change (which is the controlling factor), and none of the nullability issues is shared_preferences.dart are resolved.

rockingdice commented:
Looking forward to it!

gaetschwartz commented:
Correct, forgot to. Done now.

stuartmorgan commented on packages/shared_preferences/shared_preferences/pubspec.yaml:
```diff
@@ -5,7 +5,7 @@ homepage: https://github.com/flutter/plugins/tree/master/packages/shared_prefere
 # the version to 2.0.0.
 # See more details: https://github.com/flutter/flutter/wiki/Package-migration-to-1.0.0
-version: 0.5.13+2
+version: 0.6.0-nullsafety
```
I almost forgot (and may have missed this for other plugins): we actually need to make this 2.0.0-nullsafety. The README currently says that we guarantee 1.0 will be compatible with the current version, and encourages people setting their constraints to <2, but this is a breaking change.

(This should also update the README to remove that section.)

stuartmorgan commented on packages/shared_preferences/shared_preferences/lib/shared_preferences.dart:
```diff
@@ -54,19 +53,19 @@ class SharedPreferences {
     if (_completer == null) {
       _completer = Completer<SharedPreferences>();
```
Let's use a non-nullable local variable here, which is assigned to `_completer` at the end of the `try`, in order to remove most of the uses of `!`. `!` is a tool of last resort.

stuartmorgan commented on packages/shared_preferences/shared_preferences/lib/shared_preferences.dart:
```diff
@@ -77,79 +76,80 @@ class SharedPreferences {
   /// It is NOT guaranteed that this cache and the device prefs will remain
   /// in sync since the setter method might fail for any reason.
-  final Map<String, Object> _preferenceCache;
+  final Map<String, Object?> _preferenceCache;
```
Why are the values nullable? That's not what the platform interface that this is caching vends.

stuartmorgan commented on packages/shared_preferences/shared_preferences/lib/shared_preferences.dart:
```diff
@@ -77,79 +76,80 @@ class SharedPreferences {
 
   /// Saves a boolean [value] to persistent storage in the background.
   ///
   /// If [value] is null, this is equivalent to calling [remove()] on the [key].
```
All of the setters should take non-nullable values, and these comments removed. Nullability is a breaking change, which means we don't have to keep backwards compatibility; we should design the interface to be hard to use incorrectly in a null-safe world. In that context, we should require people to be explicit about removing values rather than accidentally passing nulls through and getting a different behavior.

(That's also consistent with `Map`, which I think is a feature.)

stuartmorgan commented on packages/shared_preferences/shared_preferences/lib/shared_preferences.dart:
```diff
@@ -77,79 +76,80 @@ class SharedPreferences {
   Future<bool> remove(String key) => _setValue(null, key, null);
 
-  Future<bool> _setValue(String valueType, String key, Object value) {
+  Future<bool> _setValue(String? valueType, String key, Object? value) {
```
With the changes mentioned above, this can take all non-nullable values, and the remove logic can move into `remove`.

stuartmorgan commented on packages/shared_preferences/shared_preferences/lib/shared_preferences.dart:
```diff
@@ -161,7 +161,7 @@ class SharedPreferences {
+      return _store.setValue(valueType!, prefixedKey, value);
     }
   }
 
```
Let's remove `commit` while making the breaking change.

gaetschwartz replied:
Not sure what you are talking about. The `commit` method is deprecated isn't it ?

stuartmorgan replied:
Right, `deprecated` means it shouldn't be used and will be removed in the future. Since it's been deprecated for a long time, and we're doing a breaking change for the plugin anway, it's a good time to actually remove it.

stuartmorgan commented on packages/shared_preferences/shared_preferences/lib/shared_preferences.dart:
```diff
@@ -181,17 +181,17 @@ class SharedPreferences {
   /// (without using the plugin) while the app is running.
   Future<void> reload() async {
-    final Map<String, Object> preferences =
+    final Map<String, Object?> preferences =
```
Same question for all of these about why the values are nullable.

stuartmorgan commented on packages/shared_preferences/shared_preferences/lib/shared_preferences.dart:
```diff
@@ -213,7 +213,8 @@ class SharedPreferences {
     SharedPreferencesStorePlatform.instance =
-        InMemorySharedPreferencesStore.withData(newValues);
+        InMemorySharedPreferencesStore.withData(
+            newValues.cast<String, Object>());
```
This cast isn't safe, since you are allowing `dynamic` in the API (meaning it won't do any type checking). You should change the API to `<String, Object>` instead of making a potentially invalid conversion at the end.

stuartmorgan commented on packages/shared_preferences/shared_preferences/test/shared_preferences_test.dart:
```diff
@@ -29,14 +27,14 @@ void main() {
 
-    FakeSharedPreferencesStore store;
-    SharedPreferences preferences;
+    FakeSharedPreferencesStore? store;
```
Why not `late`?

stuartmorgan commented on packages/shared_preferences/shared_preferences/test/shared_preferences_test.dart:
```diff
@@ -29,11 +27,11 @@ void main() {
 
     setUp(() async {
-      store = FakeSharedPreferencesStore(kTestValues);
+      store = FakeSharedPreferencesStore(kTestValues.cast<String, Object>());
```
Can't you change the declarations above rather than casting?

stuartmorgan commented on packages/shared_preferences/shared_preferences/lib/shared_preferences.dart:
```diff
@@ -83,86 +83,76 @@ class SharedPreferences {
   /// Saves a string [value] to persistent storage in the background.
-  ///
-  /// If [value] is null, this is equivalent to calling [remove()] on the [key].
+
```
Please remove.

stuartmorgan reviewed (changes requested):
One last small question, but otherwise looking good!

gaetschwartz commented:
I was also thinking about something regarding the different setters. Imagine someone sets some arbitrary data (let's say `String` here and thus could be null). The user would have to always check if null and then if it is null use `remove` if so and use `setString` if not. What about adding nullable setters as well ? Something like `setNullableString` that would itself check if null, if so use `remove` and else use `setString` ? A lot of users might probably be confused as the change to non-nullable fields in the setters are breaking and not really document how to use them. I'm pretty sure a certain amount of people used `setString(null)` to delete stuff.

What do you think ? 

I think at least updating the changelog with instructions on how to migrate something like `setString(String? s)` to 
```dart
if(s == null) prefs.remove('key');
else prefs.setString('key', s);
```

would be preferred. Thoughts ? @stuartmorgan

stuartmorgan commented:
> Imagine someone sets some arbitrary data (let's say `String` here and thus could be null).

`String` can't be null though. Do you mean `String?`? If so, the question is, what did they intend that to mean? Because `shared_preferences` *cannot* store null data for a key. If the answer is that they intended setting null to mean removing, why? That's not "set[ting] arbitrary data", it's either setting data or removing data, and they should consider why they are structuring their APIs that way.

> The user would have to always check if null and then if it is null use `remove` if so and use `setString` if not.

If they want an API that even in a NNBD world treats setting null as meaning removal, then yes. I would argue that that's poor nullsafe API design on their part, however.

> What about adding nullable setters as well ? Something like `setNullableString` that would itself check if null, if so use `remove` and else use `setString` ?

That's a badly named API though; nothing is being set in the `null` case, and in fact the opposite is happening. We certainly can't stop people from writing such wrappers if they really want them in their own code, I don't see a reason to encourage that code structure by providing them.

> A lot of users might probably be confused as the change to non-nullable fields in the setters are breaking and not really document how to use them. I'm pretty sure a certain amount of people used `setString(null)` to delete stuff.

Agreed we should call out that aspect of the breaking change in the changelog. Something like "Setters no longer accept null to mean removing values. If you were previously using `set*(key, null)` for removing, instead call `remove(key)`.

> ```dart
> void setName(String? s) async {
>   if(s == null) prefs.remove('key');
>   else prefs.setString('key', s); // Eventually prefs.setNullableString('key', s);
> }
> ```

I do not want to provide that example, because I think in most cases it's an anti-pattern for NNBD.

stuartmorgan commented on packages/shared_preferences/shared_preferences/lib/shared_preferences.dart:
```diff
@@ -83,86 +83,75 @@ class SharedPreferences {
+  }
 
   Future<bool> _setValue(String valueType, String key, Object value) {
     final String prefixedKey = '$_prefix$key';
```
The discussion about null setters made me realize that we should, as an added protection against mistakes when people are running in unsound mode, add a `ArgumentError.checkNotNull` on `value` here to make sure we never pass nulls through to the platform implementations (where the behavior would be platform-dependent).

gaetschwartz commented:
Alright, I agree, I updated the changelog and added the ArgumentError.checkNotNull(). Can you think of another example or is it okay the way I put it ?

stuartmorgan reviewed (approved):
LGTM (including README changes; thanks again for pointing out we should highlight this change)! Thanks for all the iteration on this!

I'll get it landed and published once all the CI runs are green.

gaetschwartz commented:
CI is failing for an obscure reason @stuartmorgan any idea why ? `test CHANNEL:stable` succeeded but it doesn't update https://cirrus-ci.com/task/4857579824218112.

stuartmorgan commented:
Yes, I'm not sure why it didn't update the status here. Since it did indeed pass though, we're good to land.
